File: app/src/main/java/com/hiddenramblings/tagmo/settings/Preferences.java
package com.hiddenramblings.tagmo.settings;

import android.content.Context;
import android.content.SharedPreferences;

import androidx.preference.PreferenceManager;

public class Preferences {

    private final SharedPreferences prefs;

    public Preferences(Context context) {
        prefs = PreferenceManager.getDefaultSharedPreferences(context);
    }

    private boolean getBoolean(String pref, boolean defValue) {
        return prefs.getBoolean(pref, defValue);
    }
    private void putBoolean(String pref, boolean value) {
        prefs.edit().putBoolean(pref, value).apply();
    }

    private int getInt(String pref, int defValue) {
        return prefs.getInt(pref, defValue);
    }
    private void putInt(String pref, int value) {
        prefs.edit().putInt(pref, value).apply();
    }

    private long getLong(String pref, long defValue) {
        return prefs.getLong(pref, defValue);
    }
    private void putLong(String pref, long value) {
        prefs.edit().putLong(pref, value).apply();
    }

    private String getString(String pref, String defValue) {
        return prefs.getString(pref, defValue);
    }
    private void putString(String pref, String value) {
        prefs.edit().putString(pref, value).apply();
    }

    public void remove(String pref) {
        prefs.edit().remove(pref).apply();
    }

    public String query() {
        return getString("query", null);
    }
    public void query(String value) {
        putString("query", value);
    }

    public int sort() {
        return getInt("sort", 1/*SORT.NAME*/);
    }
    public void sort(int value) {
        putInt("sort", value);
    }

    public String filterGameTitles() {
        return getString("query", null);
    }
    public void filterGameTitles(String value) {
        putString("filterGameTitles", value);
    }

    public String filterGameSeries() {
        return getString("query", null);
    }
    public void filterGameSeries(String value) {
        putString("filterGameSeries", value);
    }

    public String filterCharacter() {
        return getString("filterCharacter", null);
    }
    public void filterCharacter(String value) {
        putString("filterCharacter", value);
    }

    public String filterAmiiboSeries() {
        return getString("filterAmiiboSeries", null);
    }
    public void filterAmiiboSeries(String value) {
        putString("filterAmiiboSeries", value);
    }

    public String filterAmiiboType() {
        return getString("filterAmiiboType", null);
    }
    public void filterAmiiboType(String value) {
        putString("filterAmiiboType", value);
    }

    public int browserAmiiboView() {
        return getInt("browserAmiiboView", 1/*VIEW.COMPACT*/);
    }
    public void browserAmiiboView(int value) {
        putInt("browserAmiiboView", value);
    }

    public boolean enable_tag_type_validation() {
        return getBoolean("enable_tag_type_validation", true);
    }
    public void enable_tag_type_validation(boolean value) {
        putBoolean("enable_tag_type_validation", value);
    }

    public boolean enable_automatic_scan() {
        return getBoolean("enable_automatic_scan", true);
    }
    public void enable_automatic_scan(boolean value) {
        putBoolean("enable_automatic_scan", value);
    }

    public boolean disable_foomiibo_browser() {
        return getBoolean("disable_foomiibo_browser", false);
    }
    public void disable_foomiibo_browser(boolean value) {
        putBoolean("disable_foomiibo_browser", value);
    }

    public String image_network_settings() {
        return getString("image_network_settings", SettingsFragment.IMAGE_NETWORK_ALWAYS);
    }
    public void image_network_settings(String value) {
        putString("image_network_settings", value);
    }

    public int database_source_setting() {
        return getInt("database_source_setting", 0);
    }
    public void database_source_setting(int value) {
        putInt("database_source_setting", value);
    }

    public boolean enable_power_tag_support() {
        return getBoolean("enable_power_tag_support", false);
    }
    public void enable_power_tag_support(boolean value) {
        putBoolean("enable_power_tag_support", value);
    }

    public boolean enable_elite_support() {
        return getBoolean("enable_elite_support", false);
    }
    public void enable_elite_support(boolean value) {
        putBoolean("enable_elite_support", value);
    }

    public String settings_elite_signature() {
        return getString("settings_elite_signature", "");
    }
    public void settings_elite_signature(String value) {
        putString("settings_elite_signature", value);
    }

    public boolean enable_flask_support() {
        return getBoolean("enable_flask_support", false);
    }
    public void enable_flask_support(boolean value) {
        putBoolean("enable_flask_support", value);
    }

    public boolean settings_disable_debug() {
        return getBoolean("settings_disable_debug", false);
    }
    public void settings_disable_debug(boolean value) {
        putBoolean("settings_disable_debug", value);
    }

    public String browserRootFolder() {
        return getString("browserRootFolder", null);
    }
    public void browserRootFolder(String value) {
        putString("browserRootFolder", value);
    }

    public String browserRootDocument() {
        return getString("browserRootDocument", null);
    }
    public void browserRootDocument(String value) {
        putString("browserRootDocument", value);
    }

    public int foomiiboOffset() {
        return getInt("foomiiboOffset", -1);
    }
    public void foomiiboOffset(int value) {
        putInt("foomiiboOffset", value);
    }

    public int eliteBankCount() {
        return getInt("eliteBankCount", 200);
    }
    public void eliteBankCount(int value) {
        putInt("eliteBankCount", value);
    }

    public int eliteActiveBank() {
        return getInt("eliteActiveBank", 0);
    }
    public void eliteActiveBank(int value) {
        putInt("eliteActiveBank", value);
    }

    public int flaskActiveSlot() {
        return getInt("flaskActiveSlot", 0);
    }
    public void flaskActiveSlot(int value) {
        putInt("flaskActiveSlot", value);
    }

    public boolean recursiveFolders() {
        return getBoolean("recursiveFolders", true);
    }
    public void recursiveFolders(boolean value) {
        putBoolean("recursiveFolders", value);
    }

    public boolean preferEmulated() {
        return getBoolean("preferEmulated", false);
    }
    public void preferEmulated(boolean value) {
        putBoolean("preferEmulated", value);
    }

    public int applicationTheme() {
        return getInt("applicationTheme", 0);
    }
    public void applicationTheme(int value) {
        putInt("applicationTheme", value);
    }

    public String downloadUrl() {
        return getString("downloadUrl", null);
    }
    public void downloadUrl(String value) {
        putString("downloadUrl", value);
    }

    public String lastUpdatedAPI() {
        return getString("lastUpdatedAPI", null);
    }
    public void lastUpdatedAPI(String value) {
        putString("lastUpdatedAPI", value);
    }

    public long lastUpdatedGit() {
        return getLong("lastUpdatedGit", 0);
    }
    public void lastUpdatedGit(long value) {
        putLong("lastUpdatedGit", value);
    }
}

File: app/src/main/java/com/hiddenramblings/tagmo/Preferences.kt
package com.hiddenramblings.tagmo

import android.content.Context
import android.content.SharedPreferences
import android.net.Uri
import androidx.documentfile.provider.DocumentFile
import androidx.preference.PreferenceManager
import com.hiddenramblings.tagmo.eightbit.os.Version
import com.hiddenramblings.tagmo.eightbit.widget.FABulous

class Preferences(context: Context) {

    private val prefs: SharedPreferences

    private fun getBoolean(pref: String, defValue: Boolean): Boolean {
        return prefs.getBoolean(pref, defValue)
    }

    private fun putBoolean(pref: String, value: Boolean) {
        prefs.edit().putBoolean(pref, value).apply()
    }

    private fun getInt(pref: String, defValue: Int): Int {
        return prefs.getInt(pref, defValue)
    }

    private fun putInt(pref: String, value: Int) {
        prefs.edit().putInt(pref, value).apply()
    }

    @Suppress("SameParameterValue")
    private fun getFloat(pref: String, defValue: Float): Float {
        return prefs.getFloat(pref, defValue)
    }

    @Suppress("SameParameterValue")
    private fun putFloat(pref: String, value: Float) {
        prefs.edit().putFloat(pref, value).apply()
    }

    @Suppress("SameParameterValue")
    private fun getLong(pref: String, defValue: Long): Long {
        return prefs.getLong(pref, defValue)
    }

    @Suppress("SameParameterValue")
    private fun putLong(pref: String, value: Long) {
        prefs.edit().putLong(pref, value).apply()
    }

    private fun getString(pref: String, defValue: String?): String? {
        return prefs.getString(pref, defValue)
    }

    private fun putString(pref: String, value: String?) {
        prefs.edit().putString(pref, value).apply()
    }

    fun remove(pref: String?) {
        prefs.edit().remove(pref).apply()
    }

    private val fabulousX = "fabulousX"
    fun fabulousX(fab: FABulous): Float {
        return getFloat(fabulousX, fab.x)
    }

    fun fabulousX(value: Float) {
        putFloat(fabulousX, value)
    }

    private val fabulousY = "fabulousY"
    fun fabulousY(fab: FABulous): Float {
        return getFloat(fabulousY, fab.y)
    }

    fun fabulousY(value: Float) {
        putFloat(fabulousY, value)
    }

    private val fabulousHorzX = "fabulousHorzX"
    fun fabulousHorzX(fab: FABulous): Float {
        return getFloat(fabulousHorzX, fab.x)
    }

    fun fabulousHorzX(value: Float) {
        putFloat(fabulousHorzX, value)
    }

    private val fabulousHorzY = "fabulousHorzY"
    fun fabulousHorzY(fab: FABulous): Float {
        return getFloat(fabulousHorzY, fab.y)
    }

    fun fabulousHorzY(value: Float) {
        putFloat(fabulousHorzY, value)
    }

    private val query = "query"
    fun query(): String? {
        return getString(query, null)
    }

    fun query(value: String?) {
        putString(query, value)
    }

    private val sort = "sort"
    fun sort(): Int {
        return getInt(sort, 1 /*SORT.NAME*/)
    }

    fun sort(value: Int) {
        putInt(sort, value)
    }

    private val filterGameTitles = "filterGameTitles"
    fun filterGameTitles(): String? {
        return getString(filterGameTitles, null)
    }

    fun filterGameTitles(value: String?) {
        putString(filterGameTitles, value)
    }

    private val filterGameSeries = "filterGameSeries"
    fun filterGameSeries(): String? {
        return getString(filterGameSeries, null)
    }

    fun filterGameSeries(value: String?) {
        putString(filterGameSeries, value)
    }

    private val filterCharacter = "filterCharacter"
    fun filterCharacter(): String? {
        return getString(filterCharacter, null)
    }

    fun filterCharacter(value: String?) {
        putString(filterCharacter, value)
    }

    private val filterAmiiboSeries = "filterAmiiboSeries"
    fun filterAmiiboSeries(): String? {
        return getString(filterAmiiboSeries, null)
    }

    fun filterAmiiboSeries(value: String?) {
        putString(filterAmiiboSeries, value)
    }

    private val filterAmiiboType = "filterAmiiboType"
    fun filterAmiiboType(): String? {
        return getString(filterAmiiboType, null)
    }

    fun filterAmiiboType(value: String?) {
        putString(filterAmiiboType, value)
    }

    private val browserAmiiboView = "browserAmiiboView"
    fun browserAmiiboView(): Int {
        return getInt(browserAmiiboView, 1 /*VIEW.COMPACT*/)
    }

    fun browserAmiiboView(value: Int) {
        putInt(browserAmiiboView, value)
    }

    private val tagTypeValidation = "enable_tag_type_validation"
    fun tagTypeValidation(): Boolean {
        return getBoolean(tagTypeValidation, true)
    }

    fun tagTypeValidation(value: Boolean) {
        putBoolean(tagTypeValidation, value)
    }


    private val automaticScan = "enable_automatic_scan"
    fun automaticScan(value: Boolean) {
        putBoolean(automaticScan, value)
    }

    private val imageNetwork = "image_network_settings"
    fun imageNetwork(): String? {
        return getString(imageNetwork, GlideTagModule.IMAGE_NETWORK_ALWAYS)
    }

    fun imageNetwork(value: String?) {
        putString(imageNetwork, value)
    }

    private val databaseSource = "database_source_setting"
    fun databaseSource(): Int {
        return getInt(databaseSource, 0)
    }

    fun databaseSource(value: Int) {
        putInt(databaseSource, value)
    }

    private val powerTagEnabled = "enable_power_tag_support"
    fun powerTagEnabled(): Boolean {
        return getBoolean(powerTagEnabled, false)
    }

    fun powerTagEnabled(value: Boolean) {
        putBoolean(powerTagEnabled, value)
    }

    private val eliteEnabled = "enable_elite_support"
    fun eliteEnabled(): Boolean {
        return getBoolean(eliteEnabled, false)
    }

    fun eliteEnabled(value: Boolean) {
        putBoolean(eliteEnabled, value)
    }

    private val eliteSignature = "settings_elite_signature"
    fun eliteSignature(): String? {
        return getString(eliteSignature, "")
    }

    fun eliteSignature(value: String?) {
        putString(eliteSignature, value)
    }

    private val showCompat3DS = "settings_show_games_ds"
    fun showCompat3DS(): Boolean {
        return getBoolean(showCompat3DS, true)
    }

    fun showCompat3DS(value: Boolean) {
        putBoolean(showCompat3DS, value)
    }

    private val showCompatWiiU = "settings_show_games_wii"
    fun showCompatWiiU(): Boolean {
        return getBoolean(showCompatWiiU, true)
    }

    fun showCompatWiiU(value: Boolean) {
        putBoolean(showCompatWiiU, value)
    }

    private val showCompatSwitch = "settings_show_games_nx"
    fun showCompatSwitch(): Boolean {
        return getBoolean(showCompatSwitch, true)
    }

    fun showCompatSwitch(value: Boolean) {
        putBoolean(showCompatSwitch, value)
    }

    private val persistSkipLockInfo = "persist_skip_lock_info"
    fun persistSkipLockInfo(): Boolean {
        return getBoolean(persistSkipLockInfo, false)
    }

    fun persistSkipLockInfo(value: Boolean) {
        putBoolean(persistSkipLockInfo, value)
    }

    private val disableDebug = "settings_disable_debug"
    fun disableDebug(): Boolean {
        return getBoolean(disableDebug, false)
    }

    fun disableDebug(value: Boolean) {
        putBoolean(disableDebug, value)
    }

    private val browserRootFolder = "browserRootFolder"
    fun browserRootFolder(): String? {
        return getString(browserRootFolder, null)
    }

    fun browserRootFolder(value: String?) {
        putString(browserRootFolder, value)
    }

    private val browserRootDocument = "browserRootDocument"
    fun browserRootDocument(): String? {
        return getString(browserRootDocument, null)
    }

    fun browserRootDocument(value: String?) {
        putString(browserRootDocument, value)
    }

    private val eliteBankCount = "eliteBankCount"
    fun eliteBankCount(): Int {
        return getInt(eliteBankCount, 200)
    }

    fun eliteBankCount(value: Int) {
        putInt(eliteBankCount, value)
    }

    private val eliteActiveBank = "eliteActiveBank"
    fun eliteActiveBank(): Int {
        return getInt(eliteActiveBank, 0)
    }

    fun eliteActiveBank(value: Int) {
        putInt(eliteActiveBank, value)
    }

    private val gattActiveSlot = "gattActiveSlot"
    fun gattActiveSlot(): Int {
        return getInt(gattActiveSlot, 0)
    }

    fun gattActiveSlot(value: Int) {
        putInt(gattActiveSlot, value)
    }

    private val recursiveFolders = "recursiveFolders"
    fun recursiveFolders(): Boolean {
        return getBoolean(recursiveFolders, true)
    }

    fun recursiveFolders(value: Boolean) {
        putBoolean(recursiveFolders, value)
    }

    private val preferEmulated = "preferEmulated"
    fun preferEmulated(): Boolean {
        return getBoolean(preferEmulated, false)
    }

    fun preferEmulated(value: Boolean) {
        putBoolean(preferEmulated, value)
    }

    private val applicationTheme = "applicationTheme"
    fun applicationTheme(): Int {
        return getInt(applicationTheme, 0)
    }

    fun applicationTheme(value: Int) {
        putInt(applicationTheme, value)
    }

    val downloadUrl = "downloadUrl"
    fun downloadUrl(): String? {
        return getString(downloadUrl, null)
    }

    fun downloadUrl(value: String?) {
        putString(downloadUrl, value)
    }

    private val lastUpdatedAPI = "lastUpdatedAPI"
    fun lastUpdatedAPI(): String? {
        return getString(lastUpdatedAPI, null)
    }

    fun lastUpdatedAPI(value: String?) {
        putString(lastUpdatedAPI, value)
    }

    private val lastUpdatedGit = "lastUpdatedGit"

    fun lastUpdatedGit(): Long {
        return getLong(lastUpdatedGit, 0)
    }

    fun lastUpdatedGit(value: Long) {
        putLong(lastUpdatedGit, value)
    }

    private val lastBugReport = "lastBugReport"

    fun lastBugReport(): Long {
        return getLong(lastBugReport, 0)
    }

    fun lastBugReport(value: Long) {
        putLong(lastBugReport, value)
    }

    val isDocumentStorage : Boolean get() = run {
        Version.isLollipop && browserRootDocument()?.let {
            try {
                DocumentFile.fromTreeUri(TagMo.appContext, Uri.parse(it))
                true
            } catch (iae: IllegalArgumentException) {
                false
            }
        } ?: false
    }

    init {
        prefs = PreferenceManager.getDefaultSharedPreferences(context)
    }
}